### ISSUE TITLE
Style/RedundantParentheses should allow parenthesis for compound logical statements

### ISSUE THREAD

**Issue by postmodern** [rubocop/rubocop#12511]
Recently `rubocop` began flagging code with the [Style/RedundantParentheses](https://docs.rubocop.org/rubocop/cops_style.html#styleredundantparentheses) rule. However, it was flagging complex compound logical statements where the parenthesis help with showing the beginning and end of said statement, which helps with code readability.

```
lib/ronin/app/schemas/params_schema.rb:55:14: C: [Correctable] Style/RedundantParentheses: Don't use parentheses around a logical expression.
10
          if (param.required? && !param.has_default?)
11
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
12
lib/ronin/app/validations/nmap_params.rb:156:14: C: [Correctable] Style/RedundantParentheses: Don't use parentheses around a logical expression.
13
          if (value && !(value >= 0.0 && value <= 1.0))
14
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
15
lib/ronin/app/validations/nmap_params.rb:162:14: C: [Correctable] Style/RedundantParentheses: Don't use parentheses around a logical expression.
16
          if (value && !(value >= 0 && value <= 9))
17
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I do still believe `Style/RedundantParentheses` can be useful for catching parenthesis around singular logical statements (ex: `if (variable.nil?)`).

I believe `Style/RedundantParentheses` should make an exception for compound logical statements, or should at least be configurable to make an exception for compound logical statements. Disabling the rule entirely would allow needless parenthesis around singular logical statements to slip by.

**Comment by bbatsov** [comment]
> I do still believe Style/RedundantParentheses can be useful for catching parenthesis around singular logical statements (ex: if (variable.nil?)).

Those are super rare in practice, though, so providing special support for them seems like an overkill to me. Might be best to just implement a custom cop targeting those.

**Comment by postmodern** [comment]
@bbatsov if you check the documentation for [Style/RedundantParentheses](https://docs.rubocop.org/rubocop/cops_style.html#styleredundantparentheses) it shows:

```ruby
# bad
(x) if ((y.z).nil?)

# good
x if y.z.nil?
```

so I assume the purpose of this cop is to also eliminate parenthesis around singular statements, even though they are super rare. Despite being rare, I'd like to ensure they don't get into my code, while still allowing parenthesis around compound logical statements.

**Comment by koic** [comment]
In all the cases exemplified, Ruby style does not use parentheses around the outermost part of condition. And This cop respects parentheses that alter the precedence of the condition and targets only the outermost parentheses for removal.
Additionally, this rule is also indicated in the Ruby Style Guide, as shown here:
https://rubystyle.guide/#no-parens-around-condition

**Comment by postmodern** [comment]
@koic

> In all the cases exemplified, Ruby style does not use parentheses around the outermost part of condition.

Except the examples show only invole *singular* logical statements (ex: `if (value.nil?)`), *not* compound logical statements (ex: `if (foo && (bar || !baz))`).

> Additionally, this rule is also indicated in the Ruby Style Guide, as shown here:
https://rubystyle.guide/#no-parens-around-condition

Again, the listed examples in the Ruby Style guide only use *singular* logical statements, *not* compound logical statements. There is no authoritative style example for whether parenthesis in *compound* logical statements is allowed or prohibited This should be clarified, as currently every example only refers to singular logical statements.

**Comment by seandilda** [comment]
I would like to second this.  I've had a couple cases that prompted me to disable this cop entirely because it wanted my code to be less readable.  For instance:

```ruby
@var ||= (method1 || method2)
```

Ruby does not require the parentheses.  However, having the parentheses makes it very clear what is happening in the code.  If the parentheses are gone, I expect developers looking at the code later to have to pause and consider if the code actually works the way it is intended.

**Comment by seandilda** [comment]
This is still a concern.

**Comment by mpapis** [comment]
from rspec, trigger the cop:
```ruby
FactoryBot.define do
  factory :question do
    transient do
      answer_value { nil }
    end
    answers do
      Array((association(:answer, answer_value:) if answer_value))
    end
  end
end
```
produces:
```
spec/factories/questions.rb:7:13: C: [Correctable] Style/RedundantParentheses: Don't use parentheses around a method argument.
      Array((association(:answer, answer_value:) if answer_value))
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
addressing the cop concern:
```ruby
FactoryBot.define do
  factory :question do
    transient do
      answer_value { nil }
    end
    answers do
      Array(association(:answer, answer_value:) if answer_value)
    end
  end
end
```
produces:
```
spec/factories/questions.rb:7:49: F: Lint/Syntax: unexpected token kIF_MOD
(Using Ruby 3.3 parser; configure using TargetRubyVersion parameter, under AllCops)
      Array(association(:answer, answer_value:) if answer_value)
                                                ^^
```
So for now I have to disable the cop as it produces broken ruby code

**Comment by koic** [comment]
@mpapis FYI, this issue has been resolved in https://github.com/rubocop/rubocop/pull/14110. It hasn't been released yet.
